salt.returners.local_cache:prep_jid useless recursive function ?

arount opened the pull request:
## Issue

`salt.returners.local_cache` seems to struggle with jid directories.

There are two functions (`prep_jid` and `save_load`) trying to write in jid directories and if one of them try to add a non existing file in an existing directory (which should be an accepted behavior I think), it will recurse up to 5 time asking for new jid each time.

This may create unconsistents jids (because `prep_jid` is responsible for generating jid) and because before each recursion these functions sleeps for 0.1 seconds, it could be bad for performances.

I mostly packaged jid directory creation in func `_prep_jid_dir` and removed recursions.

salt-jenkins commented:
@arount, thanks for your PR! By analyzing the history of the files in this pull request, we identified @cachedout, @blueyed and @terminalmage to be potential reviewers.

cachedout commented:
@arount Thanks for this. Is this purpose here merely to clean up dead code and to improve readability or is there an open GitHub issue associated with this?

arount commented:
@cachedout It's about clean and make it easier to read. There is no issue related.

I will surely continue to work on this, make checks pass & update doc. I can't work all time on this, so can I ping you in comments when I'm done with this ?

In this time, can you please explain in some lines why it is done like that ? I'm afraid to miss something and create regression or something. If there are no _real_ reason why, so just tell me, I just want to be sure I really understand what I'm doing.

arount commented:
Got this with pylint: `E:  1, 0: Module file has the wrong file permissions(expected 0644): 0664 (file-perms)`. All returners are 0664, should I skip this error ?

cachedout reviewed (changes requested):
This needs a `from salt.ext.six.moves import range` at the top to pass the lint check, please.

arount committed:
returners.local_cache: remove recursions & simplify code

arount commented:
Well, I finally understood how returners works. I hope I'm not wrong. What I saw is `prep_jid` and `save_load` struggled to access jid directory.

In fact each one was trying to create jid directory but when `OSError` was raised instead of checking if directory exists, make recursions.

jids are datetime, so by waiting 0.1s for each recursion they finally got a new clean jid without existing directory.

I tested this version, hope it's ok.
btw I forced a single new commit.

arount committed:
returners.local_cache: fix pylint

arount commented:
@cachedout sorry but can you help with tests. Pylint still failling in jenkins but seems to work on my host. No sure if warning is the issue: `salt/returners/local_cache.py:28: [W0611(unused-import), ] Unused range imported from salt.ext.six.moves`.

I do not understand what going on with linode-ubuntu too.

Thank you

s0undt3ch commented:
The lint issue is valid, you do not seem to be using `range`, so, remove that import.

s0undt3ch commented:
As for the Ubuntu failures, they are unrelated to your PR, don't worry about those.

arount commented:
@s0undt3ch thank you. Done.

cachedout commented:
I know that @thatch45 has experience with some race-conditions that can be present in this area of the code so I want to run this past him before we merge it. Thanks for being patient, @arount

thatch45 reviewed (changes requested):
From the look of things I think this has the potential to create massive problems. It the event where we have jid directory collisions this would just raise the collisions instead of retry, and those collisions are much more possible in the event of a syndic.
Basically, I am very hesitant to merge this one. if we don't fall back to create a new jid then we will have issues.
I do however like the refactor and I am very open to changing the sleep to 0.01 seconds to improve performance

arount commented:
@thatch45 Thanks for feedback, this is interesting. I will look at syndic and its behavior and came back to you.

I will surely add conflict security but I do not know how. I feel like jid collision should be managed  by  `salt.utils.jid.gen_jid`. Is there something you are open to ? (not saying this is what I will do, but that can be a solution).

cachedout commented:
@thatch45 Can you follow up on the above question please? Thanks.

arount commented:
More i'm reading this more I'm thinking replacing `salt.utils.jid.gen_jid` code by `uuid.uuid4()` will fix issue without any hack or control needed.

What do you think ?

thatch45 commented:
The jid value is also a timestamp and is used elsewhere in Salt as a timestamp, I think that the best way to alleviate this issue from the perceptive of making the jid a uuid would be to add a few chars at the end of the jid that are random

cachedout commented:
@arount Are you going to come back to this one? If so, we'll need a rebase to address a merge conflict. If I don't hear from you, I'll go ahead and close it for the time being. Thanks.

arount commented:
@cachedout sorry for this delay. Yes, I want to finish it. I do all I can to push something today. Keep you in touch

cachedout commented:
@arount Thank you very much!

cachedout commented:
Hi @arount. Just checking in with you here. :]

arount commented:
Haa, i'm such a bad person ! Well I updated my branch to use `uuid` as jid (that can be transformed in timestamp) but it breaks units tests (but, it's expected), I just need to check theses tests.

So, promise, if friday I did not make it works, I will just close the PR.

I really want to finish this PR, but maybe I'm going too far.

See you soon

cachedout commented:
@arount You're great! Thanks so much!

arount committed:
Add salt.returners.local_cache ability to generate new jid in case of conflict

arount committed:
salt.utils.jid.gen_jid is now generating `uuid4` strings as jid (add tests too)

arount commented:
Ok. Here is the thing

## salt.utils.jid

The major change here is the `jid` generation and format. `salt.utils.jid.gen_jid` now returns `uuid4` strings (example `590ae156-3f01-4aa2-b603-27257ff6e627`). These uuid strings store current time and so can be parsed in timestamp or formatted string date.

### gen_jid

Because using formated `datetime.datetime.now()` allow lot of conflicts, I chose to use something more robust, uuid is done for that, so I thought the best thing to do is to use uuid and find a way to store timestamp within.

So in `salt.utils.jid.gen_jid` I now create a classic `uuid4` instance and replace the first field of this uuid by the current timestamp. Like it, even if two jid are created in the exact same time chance of conflict is really low, not impossible, but way less likely.

### jid_to_time

Jid are not returned as `uuid.UUID` instance but as string, so, to extract time from given jid `jid_to_time` will instantiate an `uuid.UUID` from given jid string and will extract its first field where timestamp is stored.

Btw I cleaned the method to compute date string by using `strftime`.

### is_jid

Just check if given string is a well formatted uuid string and check than first field is not empty or 0.

## salt.returners.local_cache

I finally add a single chance to handle conflict. I added an more explicit log that make possible to follow jid over conflicts by reading logs.

## Tests

Hm, "integration" tests file `tests/integration/returners/test_local_cache.py` was removed since I started this PR. I was struggling on this one, so I will test it anyway on my host.

I'm not sure failed tests on my host are really failing or not (maybe tests needs to be on a vm or stuff like that). I currently continue to test all that and dig in tests log to be sure, so maybe some news commits will append.

## Guys, I need you

Well, maybe I change to much thing in one single PR, don't know. But I know this can be a critic update, you know salt way better than me, I count on you to check this, if tests are ok, it should be ok, but a human point of view will make me more confident.

_wow, what a long post_

arount commented:
Hu, last thing, guys, tell me if the idea of using uuid is a good one for you, I like this solution, but maybe I'm the only one :)

arount commented:
erf, got conflict :/ Let me check that

thatch45 commented:
@arount I am sorry that you have put so much effort into this, but changing the jid format in this way will break many external salt applications, as well as many returners inside of Salt and can break many jid checks and validations. I also don't see the benefit of this PR, while the recursive approach does pin a few extra cycles it is far from being a bottle neck on the master and is only hit in the case of a collision.
Sorry, but I am going to close this out.

arount commented:
@thatch45 No worries, I understand and anyway don't have enough time to finish it like I would like.
See you on smaller PRs so :)

thatch45 commented:
Thanks for being understanding, I hate to let people down!
